src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.cs

```diff
@@ -844,5 +844,88 @@ public static IEnumerable<object[]> VerifyUploadServersStreamsAndExpectedData
+                Task<string> responseStringTask = responseTask.ContinueWith(t => t.Result.Content.ReadAsStringAsync(), TaskScheduler.Default).Unwrap();
+                Task.WaitAll(proxyTask, responseTask, responseStringTask);
+
+                TestHelper.VerifyResponseBody(responseStringTask.Result, responseTask.Result.Content.Headers.ContentMD5, false, null);
+                Assert.Equal(Encoding.ASCII.GetString(proxyTask.Result.ResponseContent), responseStringTask.Result);
+
+                NetworkCredential nc = creds != null ? creds.GetCredential(proxyUrl, "Basic") : null;
+                string expectedAuth =
+                    nc == null || nc == CredentialCache.DefaultCredentials ? null :
+                    string.IsNullOrEmpty(nc.Domain) ? $"{nc.UserName}:{nc.Password}" :
+                    $"{nc.Domain}\\{nc.UserName}:{nc.Password}";
+                Assert.Equal(expectedAuth, proxyTask.Result.AuthenticationHeaderValue);
+            }
+        }
+
+        [Theory]
+        [MemberData(nameof(BypassedProxies))]
+        public async Task Proxy_BypassTrue_GetRequestDoesntGoesThroughCustomProxy(IWebProxy proxy)
+        {
+            using (var client = new HttpClient(new HttpClientHandler() { Proxy = proxy }))
+            using (HttpResponseMessage response = await client.GetAsync(HttpTestServers.RemoteEchoServer))
+            {
+                TestHelper.VerifyResponseBody(
+                    await response.Content.ReadAsStringAsync(),
+                    response.Content.Headers.ContentMD5,
+                    false,
+                    null);
+            }
+        }
+
+        private sealed class UseSpecifiedUriWebProxy : IWebProxy
+        {
+            private readonly Uri _uri;
+            private readonly bool _bypass;
+
+            public UseSpecifiedUriWebProxy(Uri uri, ICredentials credentials = null, bool bypass = false)
+            {
+                _uri = uri;
+                _bypass = bypass;
+                Credentials = credentials;
+            }
+
+            public ICredentials Credentials { get; set; }
+            public Uri GetProxy(Uri destination) => _uri;
+            public bool IsBypassed(Uri host) => _bypass;
+        }
+
+        private sealed class PlatformNotSupportedWebProxy : IWebProxy
+        {
+            public ICredentials Credentials { get; set; }
+            public Uri GetProxy(Uri destination) { throw new PlatformNotSupportedException(); }
+            public bool IsBypassed(Uri host) { throw new PlatformNotSupportedException(); }
+        }
+
+        private static IEnumerable<object[]> BypassedProxies()
+        {
+            yield return new object[] { null };
+            yield return new object[] { new PlatformNotSupportedWebProxy() };
+            yield return new object[] { new UseSpecifiedUriWebProxy(new Uri($"http://proxydoesntexist:12345"), bypass: true) };
```

Comment:
```c#
        public static void GetRandomAuthenticatedUrl(
            bool includeDomain,
            out string authenticatedUrl,
            out NetworkCredential credential)
        {
            // Generates a random username/password to avoid the credentials being cached per process by some
            // HTTP stacks (such as WinINET) when executing multiple tests in the same process. Proper testing
            // requires that we always get 401 responses to process from the server in order to fully test
            // the authorization-related code paths.
            string user = Guid.NewGuid().ToString();
            string password = Guid.NewGuid().ToString();
            if (includeDomain)
            {
                // Use short domain name of 15 chars simulating NETBIOS length restriction.
                string domain = Guid.NewGuid().ToString().Substring(0, 15);
                credential = new NetworkCredential(user, password, domain);
                authenticatedUrl = string.Format(Resources.AuthenticatedWithDomainEchoUrlFormat, user, password, domain);
            }
            else
            {
                credential = new NetworkCredential(user, password);
                authenticatedUrl = string.Format(Resources.AuthenticatedEchoUrlFormat, user, password);
            }
        }
```